fsspec/asyn.py

```diff
@@ -92,6 +92,50 @@ def get_loop():
     return loop[0]
 
 
+try:
+    import resource
+except ImportError:
+    resource = None
+    ResourceError = OSError
+else:
+    ResourceEror = resource.error
+
+_DEFAULT_SOFT_LIMIT = 1024
+
+
+def _get_soft_limit():
+    if resource is None:
+        return _DEFAULT_SOFT_LIMIT
+
+    try:
+        soft_limit, _ = resource.getrlimit(resource.RLIMIT_NOFILE)
+    except (ImportError, ValueError, ResourceError):
+        return _DEFAULT_SOFT_LIMIT
+
+    if soft_limit == resource.RLIM_INFINITY:
+        return None
+    else:
+        return soft_limit
+
+
+async def _throttled_gather(coros, disable=False, **gather_kwargs):
+    """Run the given coroutines in smaller chunks to
+    not crossing the file descriptor limit"""
+    soft_limit = _get_soft_limit()
+
+    if disable or soft_limit is None:
+        return await asyncio.gather(*coros, **gather_kwargs)
+
+    chunk_size = soft_limit // 8
```

Comment:
Why the `// 8` here? I would expect the limit to be called the limit. 

Also, the limit is not actually settable (unless it falls back to 1024), but I would expect it to be, probably have an option in the config too.

Reply:
> Why the // 8 here? I would expect the limit to be called the limit.

we might increase it to `// 4` though I don't think it would be safe to assume all file descripts of this process can be allocated in this procedure alone. Then it would fail if even a couple other files are open

> Also, the limit is not actually settable (unless it falls back to 1024), but I would expect it to be, probably have an option in the config too.

It can be set using `uname` though I guess we could expose it with `put()`.

Reply:
Sorry, you misunderstood what I wanted to say

> we might increase it to // 4

I mean, that we should not have `//` by any factor here, and call the variable "gather batch size" from the start. In the function above where it is determined, we can have `// 8`, but the batch size can be changed by the user.

> It can be set using uname

I don't mean changing the actual OS limit, that seem not like something we should ver do. Rather, that we can guess a reasonable batchsize, but the user can change this batch size if they want.

Reply:
> I don't mean changing the actual OS limit, that seem not like something we should ver do. Rather, that we can guess a reasonable batchsize, but the user can change this batch size if they want.

Makes sense.